File: Exercise_1.java
/*
 * Time Complexity: 
 * 1. isEmpty - O(1)
 * 2. push - O(1)
 * 3. pop - O(1)
 * 4. peek - O(1)
 * 
 * Space complexity - O(MAX)
 */
class Stack { 
    //Please read sample.java file before starting.
  //Kindly include Time and Space complexity at top of each file
    static final int MAX = 1000; 
    int top; 
    int a[] = new int[MAX]; // Maximum size of Stack 
  
    boolean isEmpty() 
    { 
        return top < 0;
    } 

    Stack() 
    { 
        //Initialize your constructor 
        top = -1;
    } 
  
    boolean push(int x) 
    { 
        //Check for stack Overflow
        if(top == MAX) {
            return false;
        }
        a[++top] = x;
        return true;
        
    } 
  
    int pop() 
    { 
        //If empty return 0 and print " Stack Underflow"
        if(top == -1) {
            System.out.println("Stack Underflow");
            return 0;
        }
        return a[top--];
        //Write your code here
    } 
  
    int peek() 
    { 
        return a[top];
    } 
} 
  
// Driver code 
class Main { 
    public static void main(String args[]) 
    { 
        Stack s = new Stack(); 
        s.push(10); 
        s.push(20); 
        s.push(30); 
        System.out.println(s.pop() + " Popped from stack"); 
        System.out.println("Peek: " + s.peek());
        System.out.println("isEmpty: " + s.isEmpty());

    } 
}

File: Exercise_1.cpp
#include <iostream>

using namespace std;
 
#define MAX 1000
#define EMPTY -9999
 
class Stack {
  //Please read sample.java file before starting.
  //Kindly include Time and Space complexity at top of each file
    int top;
 
public:
    int a[MAX]; // Maximum size of Stack
    Stack() {
        int top = -1;
    }
    bool push(int x);
    int pop();
    int peek();
    bool isEmpty();
};
 
bool Stack::isEmpty()
{
    return (-1 == top);
}

bool Stack::push(int x)
{
    //Your code here
    //Check Stack overflow as well
    if (top >= MAX)
        return false;
    a[++top] = x;
   
    return true;
}
 
int Stack::pop()
{
    //Your code here
    //Check Stack Underflow as well
    if (isEmpty())
        return EMPTY;
    
    return a[top--];
}

int Stack::peek()
{
    //Your code here
    //Check empty condition too
    if (isEmpty())
        return EMPTY;
    return a[top];
}
 
// Driver program to test above functions
int main()
{
    class Stack s;
    // s.push(10);
    // s.push(20);
    // s.push(30);
    cout << s.pop() << " Popped from stack\n";
 
    return 0;
}